Creates a Rating Widget React Component

rsahlstrom opened the pull request:
Adds a react rating widget component that allows you to specify a size,
    control hover state, assign an initial rating, supports half stars, and
    can be disabled. Component also has a onRate callback that is called
    when a rating is assigned.

kbratko committed:
Remove mistaken added task

kbratko committed:
Create rating stars first variant

kbratko committed:
Create rating stars second variant

Commit:
Creates a Rating Widget React Component

Adds a react rating widget component that allows you to specify a size,
control hover state, assign an initial rating, supports half stars, and
can be disabled. Component also has a onRate callback that is called
when a rating is assigned.

rsahlstrom commented on src/components/RatingWidget.jsx:
```diff
@@ -0,0 +1,111 @@
+    initialRating: React.PropTypes.number,
+    halfRating: React.PropTypes.bool,
+    hover: React.PropTypes.bool,
+  },
```
Here we're defining our propTypes so we can enforce what data we're expecting and can alert our implementers on what errors they're making.

rsahlstrom commented on src/components/RatingWidget.jsx:
```diff
@@ -0,0 +1,111 @@
+    return {
+      rating: this.props.initialRating,
+      tempRating: null
+    }
```
We use state to keep track of what the current rating is and what the mouseOver rating is.

rsahlstrom commented on src/components/RatingWidget.jsx:
```diff
@@ -0,0 +1,111 @@
+    this.setState({rating: newRating});
+    this.setState({tempRating: null});
+    this.props.onRate(newRating);
+  },
```
All handleClick does is set the state. We do no DOM manipulation.

jehoshua02 replied:
I like to put it this way:

If a method modifies state, modifying state is the only thing it does.

This goes along with the [SOLID Single Responsibility Principle](http://en.wikipedia.org/wiki/SOLID_%28object-oriented_design). I like functions that only do one thing.

rsahlstrom commented on src/components/RatingWidget.jsx:
```diff
@@ -0,0 +1,111 @@
+    this.setState({tempRating: null});
+  },
+
+  calchalfRating: function(newRating, e) {
```
I should probably rename this function to calcHalfRating. I'll fix that.

rsahlstrom commented on src/components/RatingWidget.jsx:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    newTempRating = this.calchalfRating(newTempRating, e);
+    this.setState({tempRating: newTempRating})
```
As above, hover only worries about setting state.

rsahlstrom commented on src/components/RatingWidget.jsx:
```diff
@@ -0,0 +1,111 @@
+        'ddm-rating-widget__step--whole': showWhole,
+        'ddm-rating-widget__step--half': showHalf,
+        'ddm-rating-widget__step--hover': this.state.tempRating
+      }
```
This is where the magic happens. We look at the state to determine which classes to add or remove from the step classes.

rsahlstrom commented:
@kbratko Please review this and your other pull requests and then take it apart to see why we did what. Let me know any questions you have.

kbratko commented:
After reviewing the code I would like to add some additional functionality for the best ui.
1.  Add cursor:pointer property to css class .ddm-rating-widget__step to show user that the rating can be modified. Also it would be better to add .ddm-rating-widget__step--disabled class when rating disabled property is true. And set up css property for this class cursor:default to show user that this rating can't be modified.
2. When I checked how rating works on my iPhone simulator I noticed that the stars is too small for being clicked correctly. It will be so difficult to tap them with fat fingers.
   So I decided to increase stars size. To change star size means that I should change .ddm-rating-widget__step font-size property. But in this case I should also change the height and width of it, which should be equal to the font-size (in case width  property has not correct value, the half star has wrong view). Also adding the line-height property, which also equal the font-size, allows to see the perfect square area around the star. 

So as we use sass (thats so great!) we can do it without any effort:

$stepSize: 30px;

.ddm-rating-widget__step {
  font-size: $stepSize;
  width: $stepSize;
  height: $stepSize;
  line-height: $stepSize;
}

or use mixin (make styles more reusable):

@mixin size($stepSize) {
  font-size: $stepSize;
  width: $stepSize;
  height: $stepSize;
  line-height: $stepSize;
}

.ddm-rating-widget__step { @include size(30px); }

It allows developer to change size in the only one place and all another properties can be set up automaticaly.

kbratko commented:
One more thing. When I tap stars on iPhone the stars become blue and don't become orange when rate is done. As I know hover event doesn't work on mobile. Can we left orange color for both hover and whole states?

Commit:
Adds Larger Size for Small Screens, Disable/Enable, Style Updates

Adds a media query to make the rating steps larger on small screens
to make the click areas easier to hit on mobile devices.

Adds methods to enable and disable the ratings widget and adds examples
of how this functionality can be used on the examples page.

Disable the mouse move events on devices with touch events to avoid
requiring a double click on a mobile phones to select a rating.

Updates the README.md and adds a ghpages tasks to publish examples to
GitHub pages.

rsahlstrom commented:
@kbratko Good feedback. I've pushed another commit with updated changes. Will you please take a look again?

@jspencerDDM Made some changes and got the disabled stuff working I tried yesterday. You may be interested in seeing this second commit.
